linera-service/src/config.rs

```diff
@@ -210,41 +226,82 @@ impl WalletState {
-        self.genesis_config.admin_id
+        self.inner.genesis_config.admin_id
     }
 
     pub fn genesis_config(&self) -> &GenesisConfig {
-        &self.genesis_config
+        &self.inner.genesis_config
     }
 
-    pub fn read(path: &Path) -> Result<Self, anyhow::Error> {
-        let file = OpenOptions::new().read(true).open(path)?;
-        Ok(serde_json::from_reader(BufReader::new(file))?)
+    pub fn from_file(path: &Path) -> Result<Self, anyhow::Error> {
+        let file = FileOptions::new().read(true).write(true);
+        let block = false;
+        let file_lock = match FileLock::lock(path, block, file) {
+            Ok(lock) => lock,
+            Err(err) => bail!("Error getting write lock to wallet: {}", err),
+        };
+        let inner = serde_json::from_reader(BufReader::new(&file_lock.file))?;
+        Ok(Self {
+            inner,
+            wallet_path: path.into(),
+            _lock: file_lock,
+        })
     }
 
     pub fn create(path: &Path, genesis_config: GenesisConfig) -> Result<Self, anyhow::Error> {
-        let file = OpenOptions::new()
-            .create(true)
-            .write(true)
-            .read(true)
-            .open(path)?;
-        let mut reader = BufReader::new(file);
+        let file = FileOptions::new().create(true).write(true).read(true);
+        let block = false;
+        let file_lock = match FileLock::lock(path, block, file) {
+            Ok(lock) => lock,
+            Err(err) => bail!("Error getting write lock to wallet: {}", err),
+        };
+        let mut reader = BufReader::new(&file_lock.file);
         if reader.fill_buf()?.is_empty() {
-            return Ok(Self {
+            let inner = InnerWallet {
                 chains: Default::default(),
                 unassigned_key_pairs: Default::default(),
                 default: None,
                 genesis_config,
-            });
+            };
+            Ok(Self {
+                inner,
+                wallet_path: path.into(),
+                _lock: file_lock,
+            })
+        } else {
+            let inner = serde_json::from_reader(reader)?;
+            Ok(Self {
+                inner,
+                wallet_path: path.into(),
+                _lock: file_lock,
+            })
         }
-        Ok(serde_json::from_reader(reader)?)
     }
 
-    pub fn write(&self, path: &Path) -> Result<(), std::io::Error> {
-        let file = OpenOptions::new().write(true).open(path)?;
-        let mut writer = BufWriter::new(file);
-        serde_json::to_writer_pretty(&mut writer, &self)?;
+    /// Writes the wallet to disk.
+    ///
+    /// The contents of the wallet need to be over-written completely, so
+    /// a temporary file is created as a backup in case a crash occurs while
+    /// writing to disk.
+    ///
+    /// The temporary file is then deleted once we successfully save the
+    /// contents of the wallet.
+    pub fn write(&mut self) -> Result<(), std::io::Error> {
```

Comment:
So the usual way to do this is actually simpler:
* write the new data to a temporary file starting with the same name such as `{file-name}.{random}` 
* in case of success, rename the temporary file to the actual file name (this operation is guaranteed to be atomic)
* otherwise, (try to) remove the temporary file before exiting

Reply:
That's a good idea!